File: apps/api/app/models/training.py
from datetime import datetime
from enum import StrEnum

from sqlalchemy import Enum, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, IdMixin, TimestampMixin


class TrainingStatus(StrEnum):
    open = "open"
    upcoming = "upcoming"


class TrainingFormat(StrEnum):
    online = "online"
    in_person = "in_person"


class Training(IdMixin, TimestampMixin, Base):
    __tablename__ = "trainings"

    slug: Mapped[str] = mapped_column(String(200), unique=True, index=True)
    title: Mapped[str] = mapped_column(String(300))
    status: Mapped[TrainingStatus] = mapped_column(Enum(TrainingStatus, native_enum=False, length=20))
    format: Mapped[TrainingFormat] = mapped_column(Enum(TrainingFormat, native_enum=False, length=20))
    starts_at: Mapped[datetime | None]
    ends_at: Mapped[datetime | None]
    expected_label: Mapped[str] = mapped_column(String(100), default="")
    trainers: Mapped[str] = mapped_column(String(500), default="")
    description: Mapped[str] = mapped_column(Text, default="")
    objectives: Mapped[str] = mapped_column(Text, default="")
    agenda: Mapped[str] = mapped_column(Text, default="")
    registration_url: Mapped[str | None] = mapped_column(String(1000))
    cover_image_url: Mapped[str | None] = mapped_column(String(1000))

File: apps/web/src/lib/training.ts
import type { Training, TrainingInput } from '@/api/types'
import type { TrainingValues } from './schemas'

export const slugify = (text: string) =>
  text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)/g, '')

export const emptyTrainingValues: TrainingValues = {
  title: '',
  slug: '',
  status: 'open',
  format: 'online',
  date: '',
  start_time: '09:00',
  end_time: '16:00',
  expected_label: '',
  trainers: '',
  description: '',
  objectives: '',
  agenda: '',
  registration_url: '',
  cover_image_url: null,
}

export const toTrainingValues = (t: Training): TrainingValues => ({
  title: t.title,
  slug: t.slug,
  status: t.status,
  format: t.format,
  date: t.starts_at?.slice(0, 10) ?? '',
  start_time: t.starts_at?.slice(11, 16) ?? '',
  end_time: t.ends_at?.slice(11, 16) ?? '',
  expected_label: t.expected_label,
  trainers: t.trainers,
  description: t.description,
  objectives: t.objectives,
  agenda: t.agenda,
  registration_url: t.registration_url ?? '',
  cover_image_url: t.cover_image_url,
})

export const toTrainingInput = (v: TrainingValues): TrainingInput => ({
  title: v.title,
  slug: v.slug,
  status: v.status,
  format: v.format,
  starts_at: v.date ? `${v.date}T${v.start_time || '00:00'}:00` : null,
  ends_at: v.date && v.end_time ? `${v.date}T${v.end_time}:00` : null,
  expected_label: v.expected_label,
  trainers: v.trainers,
  description: v.description,
  objectives: v.objectives,
  agenda: v.agenda,
  registration_url: v.registration_url || null,
  cover_image_url: v.cover_image_url,
})
